Added automatic subsequence selection, added range parameter for larg…

martin-mann opened the pull request:
…e input files (reference)

martin-mann commented on src/IntaRNA/Accessibility.cpp:
```diff
@@ -114,4 +114,57 @@ writeRNAplfold_text( std::ostream& out, const E_type RT, const bool writeProbs )
 
+void
+Accessibility::
+compute_accessible_ranges( IndexRangeList& ranges, const size_t max_seq_length, E_type threshold, int i, int j) const
```
Hi @agawrons, just to ensure I get the idea right:

this function will generate all ranges that have an ED below `threshold`, which also holds for all covered subranges.

right?

martin-mann commented on src/bin/IntaRNA.cpp:
```diff
@@ -154,9 +157,31 @@ int main(int argc, char **argv){
+									targetRegions = parameters.getTargetRanges(targetNumber);
+								}
+								else{
+									targetAcc->compute_accessible_ranges( targetRegions, tRegionLenMax, Accessibility::ED_UPPER_BOUND, 0, targetAcc->getSequence().size());
```
seems the `threshold` is the ED upper bound so far and thus only filters for non-accessible sites right?

martin-mann commented on src/bin/CommandLineParsing.cpp:
```diff
@@ -982,6 +1018,51 @@ parseRegion( const std::string & argName, const std::string & value, const RnaSe
+	} else
+	// no valid input
+	{
+		LOG(ERROR) <<"the argument for "<<argName<<" is neither a valid range string encoding nor a file that can be found";
```
you do no file handling here.. so i think the error message wrong, or?

martin-mann commented on src/bin/IntaRNA.cpp:
```diff
@@ -96,7 +99,7 @@ int main(int argc, char **argv){
 		# pragma omp parallel for schedule(dynamic) num_threads( parameters.getThreads() ) shared(queryAcc,reportedInteractions,exceptionPtrDuringOmp,exceptionInfoDuringOmp) if(parallelizeTargetLoop)
 #endif
-		for ( size_t targetNumber = 0; targetNumber < parameters.getTargetSequences().size(); ++targetNumber )
+		for ( size_t targetNumber = target_range.from; targetNumber <= target_range.to; ++targetNumber )
```
this disables multi-sequence support doesnt it?

martin-mann replied:
this only supports a single range so far... if you want to enable multiple ranges, you need an additional `for`-loop..

martin-mann commented:
Hi @agawrons , I am not sure if I get it right, as far as I understand you want to do the following:
- given an upper bound on ED and a maximal interaction length, which define 'accessible regions'
- restrict prediction to these 'accessible regions'

right?

agawrons commented:
>
> this function will generate all ranges that have an ED below threshold,
> which also holds for all covered subranges.
>
> right?
>
It will generate ranges less than max_seq_length by splitting at regions
with the highest ED. Threshold changes at each recursive call and is the
max ED of the previous recursion (some subsequence longer than
max_seq_length). ED at each position is determined by a sliding window.

seems the threshold is the ED upper bound so far and thus only filters for
> non-accessible sites right?
>

For the first call it is set to ED upper bound to determine the max ED
within the entire sequence. When it recurses there will be at least one
position equal to the threshold and the sequence will be split. I initially
used E_type max() but it seemed like this value was more appropriate. Now
I'm not sure... is it a user defined variable?

you do no file handling here.. so i think the error message wrong, or?
>

Yes, this is wrong. Copy and paste error.

this disables multi-sequence support doesnt it?
>

This is precisely for multiple sequence input. With this parameter you can
pass the reference transcriptome and only run computation of a subset of
sequences. For my application I used this to avoid splitting the reference
into 1000 or more separate files. This seemed like a more elegant solution.
It could be implemented better... this way the entire file is still read.
It should be implemented at the point the file is read.

Hi @agawrons <https://github.com/agawrons> , I am not sure if I get it
> right, as far as I understand you want to do the following:
>
>    - given an upper bound on ED and a maximal interaction length, which
>    define 'accessible regions'
>    - restrict prediction to these 'accessible regions'
>
> right?
>
Almost. I think me using ED upper bound rather than E_type max() leads to
confusion. The user only specifies the length and the algorithm will
progressively reduce the ED threshold until only a set of ranges below the
length threshold exist. Perhaps there should be a separate public function
that calls the recursive function so nobody accidentally sets these values
(ED, start, end).

I'll quickly make these changes and push. Give me an hour or so.

Alex

On Fri, Nov 3, 2017 at 7:00 AM, Martin Mann <notifications@github.com>
wrote:

> Hi @agawrons <https://github.com/agawrons> , I am not sure if I get it
> right, as far as I understand you want to do the following:
>
>    - given an upper bound on ED and a maximal interaction length, which
>    define 'accessible regions'
>    - restrict prediction to these 'accessible regions'
>
> right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BackofenLab/IntaRNA/pull/95#issuecomment-341710872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbAQ8ny11dXnohBHP3medhRLPmGrGCgnks5syxxfgaJpZM4QRG1n>
> .
>

agawrons commented:
I see UPPER_BOUND is E_INF, so is it any different than E_Type max()? I
guess the former is greater than the latter,

On Fri, Nov 3, 2017 at 11:08 AM, Alex Gawronski <alex.gawronski@gmail.com>
wrote:

> this function will generate all ranges that have an ED below threshold,
>> which also holds for all covered subranges.
>>
>> right?
>>
> It will generate ranges less than max_seq_length by splitting at regions
> with the highest ED. Threshold changes at each recursive call and is the
> max ED of the previous recursion (some subsequence longer than
> max_seq_length). ED at each position is determined by a sliding window.
>
> seems the threshold is the ED upper bound so far and thus only filters
>> for non-accessible sites right?
>>
>
> For the first call it is set to ED upper bound to determine the max ED
> within the entire sequence. When it recurses there will be at least one
> position equal to the threshold and the sequence will be split. I initially
> used E_type max() but it seemed like this value was more appropriate. Now
> I'm not sure... is it a user defined variable?
>
> you do no file handling here.. so i think the error message wrong, or?
>>
>
> Yes, this is wrong. Copy and paste error.
>
> this disables multi-sequence support doesnt it?
>>
>
> This is precisely for multiple sequence input. With this parameter you can
> pass the reference transcriptome and only run computation of a subset of
> sequences. For my application I used this to avoid splitting the reference
> into 1000 or more separate files. This seemed like a more elegant solution.
> It could be implemented better... this way the entire file is still read.
> It should be implemented at the point the file is read.
>
> Hi @agawrons <https://github.com/agawrons> , I am not sure if I get it
>> right, as far as I understand you want to do the following:
>>
>>    - given an upper bound on ED and a maximal interaction length, which
>>    define 'accessible regions'
>>    - restrict prediction to these 'accessible regions'
>>
>> right?
>>
> Almost. I think me using ED upper bound rather than E_type max() leads to
> confusion. The user only specifies the length and the algorithm will
> progressively reduce the ED threshold until only a set of ranges below the
> length threshold exist. Perhaps there should be a separate public function
> that calls the recursive function so nobody accidentally sets these values
> (ED, start, end).
>
> I'll quickly make these changes and push. Give me an hour or so.
>
> Alex
>
> On Fri, Nov 3, 2017 at 7:00 AM, Martin Mann <notifications@github.com>
> wrote:
>
>> Hi @agawrons <https://github.com/agawrons> , I am not sure if I get it
>> right, as far as I understand you want to do the following:
>>
>>    - given an upper bound on ED and a maximal interaction length, which
>>    define 'accessible regions'
>>    - restrict prediction to these 'accessible regions'
>>
>> right?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/BackofenLab/IntaRNA/pull/95#issuecomment-341710872>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AbAQ8ny11dXnohBHP3medhRLPmGrGCgnks5syxxfgaJpZM4QRG1n>
>> .
>>
>
>

martin-mann commented:
@agawrons you are hunting for the most accessible regions below a given length-threshold, right?

martin-mann commented on src/bin/CommandLineParsing.cpp:
```diff
@@ -203,6 +209,19 @@ CommandLineParsing::CommandLineParsing()
+		("qRange"
+			, value<std::string>(&(qRangeString))
+				->notifier(boost::bind(&CommandLineParsing::validate_qRange,this,_1))
+			, std::string("query selection : the range of sequences to use in the provided FASTA file (Useful when using entire transcriptome as input). In the format 'from1-to1' assuming indexing starts with 0").c_str())
```
user interface indexing should always start with "1" rather than "0"..

agawrons commented:
I attached a figure. The straight line is the very long RNA sequence and
curve is the accessibility energy at any given position (actually any
window at a given position to smooth the curve). The algorithm splits the
sequence at the point of highest energy. If the subsequences are still too
long, it is split again. This continues until all subsequences are below
the length threshold (the seven leaves in this case). The starts and ends
of these subsequences are the regions used for prediction. It's just a
better way of splitting long sequences than using fixed size overlapping
windows. My assumption is that the optimal RNA-RNA interactions will not
fall in the regions of highest accessibility energy (most structured). Even
if they did, I'm not sure if they would truly occur in nature.

this only supports a single range so far... if you want to enable multiple
> ranges, you need an additional for-loop..


Do you think anyone would need to specify multiple ranges?  I only added
this so I could easily distribute the work on a cluster (one range for each
job). Just to be clear on nomenclature, I call positions on the sequences a
"region" and a set of sequences a "range". So you can still have multiple
regions for each sequence in a range.
I wasn't going to add this to the code of the branch initially since it was
for my specific purpose but then I thought maybe someone would find it
useful. You don't need to include it if you don't want. I could just refer
people to my branch for this project. I don't want to go back to manually
splitting files since it's a pain.

user interface indexing should always start with "1" rather than "0"..


Ok, I can change this. I chose zero based since that's what my other code
uses. Since these are not positions on the sequences I didn't think I
needed to follow the convention.

On Mon, Nov 6, 2017 at 1:25 AM, Martin Raden <notifications@github.com>
wrote:

> *@martin-mann* commented on this pull request.
> ------------------------------
>
> In src/bin/CommandLineParsing.cpp
> <https://github.com/BackofenLab/IntaRNA/pull/95#discussion_r149027047>:
>
> > @@ -203,6 +209,19 @@ CommandLineParsing::CommandLineParsing()
>  			, value<std::string>(&(qRegionString))
>  				->notifier(boost::bind(&CommandLineParsing::validate_qRegion,this,_1))
>  			, std::string("interaction site : query regions to be considered for interaction prediction. Either given as BED file (for multi-sequence FASTA input) or in the format 'from1-to1,from2-to2,..' assuming indexing starts with 1").c_str())
> +		("qRegionLenMax"
> +			, value<int>(&(qRegionLenMax.val))
> +				->default_value(qRegionLenMax.def)
> +				->notifier(boost::bind(&CommandLineParsing::validate_qRegionLenMax,this,_1))
> +			, std::string("interaction site : automatically select target regions to be considered for interaction prediction"
> +					" up to the specified length based on accessibility. Overrides --qRegion."
> +					" (arg in range ["+toString(qRegionLenMax.min)+",QUERY_LEN];"
> +					" Disabled by default (value of 0))"
> +					).c_str())
> +		("qRange"
> +			, value<std::string>(&(qRangeString))
> +				->notifier(boost::bind(&CommandLineParsing::validate_qRange,this,_1))
> +			, std::string("query selection : the range of sequences to use in the provided FASTA file (Useful when using entire transcriptome as input). In the format 'from1-to1' assuming indexing starts with 0").c_str())
>
> user interface indexing should always start with "1" rather than "0"..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BackofenLab/IntaRNA/pull/95#pullrequestreview-74356549>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbAQ8hKXPPOqRGXsGZZEzaG7UCO8sM6Zks5sztCEgaJpZM4QRG1n>
> .
>

martin-mann commented on src/IntaRNA/Accessibility.cpp:
```diff
@@ -114,4 +114,71 @@ writeRNAplfold_text( std::ostream& out, const E_type RT, const bool writeProbs )
+	int to = -1;
+	E_type ED = ED_UPPER_BOUND;
+
+	for (i; i<j+1-l; i++) {
```
this setup prevents splits eg at `i+l` if `j=i+(1.5*l)`, sinc i runs only up to `i+(0.5*l)`

ie, you do not consider all valid splits.. 

if not interested, you can also initialize the running i with initiali+l

martin-mann commented on src/IntaRNA/Accessibility.cpp:
```diff
@@ -153,7 +158,7 @@ compute_accessible_ranges_recursive( IndexRangeList& ranges, const size_t max_se
 				}
 				else{
-					ranges.push_back(IndexRange(from+1,to+1));
+					ranges.push_back(IndexRange(from,to));
```
huuh.. the index-start-1-convention only applies to user input! internally you should use index-start-0 of course.. uaks.. sorry for the confusion!

agawrons commented:
>
> this setup prevents splits eg at i+l if j=i+(1.5*l), sinc i runs only up
> to i+(0.5*l)
>
> ie, you do not consider all valid splits..
>
> if not interested, you can also initialize the running i with initiali+l
>
I guess I should enforce that max_seq_len >= l, then it would not matter
since you would never split at i+l if if j=i+(1.5*l) (i+0.5*l would make a
region sufficiently short). In other words, since I check all windows of
size l, I would check all possible split points, if all split points must
produce a region of length > l.

your whole ED check setup is based on the arbitrary accessibility window
> length of 50.
>
> thus, shorter interactions (that might have a much lower ED) are hidden
> and might be 'split away'
>
> or do i miss something?
>
> if i am right, wouldnt it be better to identify the minimal ED of any
> window per position and to use these values to cut out unaccessible sites?
>
> did you test l=1 for position-wise ED checks?
>
 This could happen, but your short accessible region (<50) would need to be
flanked by very inaccessible regions to make this window "beat" other
windows of inaccessible regions for selection as a split point. However I'm
not happy with this and I rather not have an arbitrary selection for this.
I guess it could make sense to use the seed length maybe? But then it would
only apply in cases where the user uses a seed.
I can test l=1. When I think about it, it should be ok.

huuh.. the index-start-1-convention only applies to user input! internally
> you should use index-start-0 of course.. uaks.. sorry for the confusion!
>

I think when I first implemented this I saw what the user needs to specify
as regions (1-based) and made my code produce 1-based regions. I didn't
notice your conversion to 0-based at the validation stage. So after this
change the regions will be the correct 0-based values. In my tests I indeed
see that this is the correct way. I actually see interactions at 1--n, when
before I saw only 2--n *facepalm*. All my analysis will be wrong for these
cases, but I think in the grand scheme of things 1nt won't really matter.

On Tue, Nov 7, 2017 at 3:01 AM, Martin Raden <notifications@github.com>
wrote:

> *@martin-mann* commented on this pull request.
> ------------------------------
>
> In src/IntaRNA/Accessibility.cpp
> <https://github.com/BackofenLab/IntaRNA/pull/95#discussion_r149334166>:
>
> > @@ -153,7 +158,7 @@ compute_accessible_ranges_recursive( IndexRangeList& ranges, const size_t max_se
>  					compute_accessible_ranges_recursive(ranges, max_seq_length, ED, from, to);
>  				}
>  				else{
> -					ranges.push_back(IndexRange(from+1,to+1));
> +					ranges.push_back(IndexRange(from,to));
>
> huuh.. the index-start-1-convention only applies to user input! internally
> you should use index-start-0 of course.. uaks.. sorry for the confusion!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BackofenLab/IntaRNA/pull/95#pullrequestreview-74709159>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbAQ8ncmlihReexG3NRj0voi7YfbrEJnks5s0DhvgaJpZM4QRG1n>
> .
>

agawrons commented:
l=2 is the lowest I could go without modifying the code. The resulting
regions for BRCA2 are below. They are fairly similar; it appears using a
larger window cuts out more sequence in between regions. A short region at
89-99 is lost as you predicted. But would there be a meaningful interaction
in such a small region? In my analysis nothing of that size would pass my
filtering. In this case the optimal interaction at that location is -4.9
kcal/mol.

I wonder if I should enforce some minimum region length to avoid creating
these short regions. Or throw them out since they are likely fairly
structured regions.

Do you have good data for testing where the "truth" is known? When it comes
to lncRNA almost nothing is known.

l=2
# INFO  : 0-32
# INFO  : 33-88
# INFO  : 89-99
# INFO  : 101-337
# INFO  : 338-1791
# INFO  : 1792-2439
# INFO  : 2440-3021
# INFO  : 3022-3222
# INFO  : 3223-4208
# INFO  : 4209-5048
# INFO  : 5051-6053
# INFO  : 6054-7067
# INFO  : 7068-8037
# INFO  : 8039-8291
# INFO  : 8292-8504
# INFO  : 8505-9860
# INFO  : 9861-10737
# INFO  : 10738-11386

l=50
# INFO  : 0-62
# INFO  : 117-321
# INFO  : 322-1435
# INFO  : 1436-2162
# INFO  : 2165-3332
# INFO  : 3334-4746
# INFO  : 4750-6072
# INFO  : 6073-6411
# INFO  : 6415-7817
# INFO  : 7826-8491
# INFO  : 8500-8668
# INFO  : 8669-8944
# INFO  : 8945-9463
# INFO  : 9467-10646
# INFO  : 10655-10799
# INFO  : 10801-11386

On Tue, Nov 7, 2017 at 10:59 AM, Alex Gawronski <alex.gawronski@gmail.com>
wrote:

> this setup prevents splits eg at i+l if j=i+(1.5*l), sinc i runs only up
>> to i+(0.5*l)
>>
>> ie, you do not consider all valid splits..
>>
>> if not interested, you can also initialize the running i with initiali+l
>>
> I guess I should enforce that max_seq_len >= l, then it would not matter
> since you would never split at i+l if if j=i+(1.5*l) (i+0.5*l would make
> a region sufficiently short). In other words, since I check all windows of
> size l, I would check all possible split points, if all split points must
> produce a region of length > l.
>
> your whole ED check setup is based on the arbitrary accessibility window
>> length of 50.
>>
>> thus, shorter interactions (that might have a much lower ED) are hidden
>> and might be 'split away'
>>
>> or do i miss something?
>>
>> if i am right, wouldnt it be better to identify the minimal ED of any
>> window per position and to use these values to cut out unaccessible sites?
>>
>> did you test l=1 for position-wise ED checks?
>>
>  This could happen, but your short accessible region (<50) would need to
> be flanked by very inaccessible regions to make this window "beat" other
> windows of inaccessible regions for selection as a split point. However I'm
> not happy with this and I rather not have an arbitrary selection for this.
> I guess it could make sense to use the seed length maybe? But then it would
> only apply in cases where the user uses a seed.
> I can test l=1. When I think about it, it should be ok.
>
> huuh.. the index-start-1-convention only applies to user input! internally
>> you should use index-start-0 of course.. uaks.. sorry for the confusion!
>>
>
> I think when I first implemented this I saw what the user needs to specify
> as regions (1-based) and made my code produce 1-based regions. I didn't
> notice your conversion to 0-based at the validation stage. So after this
> change the regions will be the correct 0-based values. In my tests I indeed
> see that this is the correct way. I actually see interactions at 1--n, when
> before I saw only 2--n *facepalm*. All my analysis will be wrong for these
> cases, but I think in the grand scheme of things 1nt won't really matter.
>
> On Tue, Nov 7, 2017 at 3:01 AM, Martin Raden <notifications@github.com>
> wrote:
>
>> *@martin-mann* commented on this pull request.
>> ------------------------------
>>
>> In src/IntaRNA/Accessibility.cpp
>> <https://github.com/BackofenLab/IntaRNA/pull/95#discussion_r149334166>:
>>
>> > @@ -153,7 +158,7 @@ compute_accessible_ranges_recursive( IndexRangeList& ranges, const size_t max_se
>>  					compute_accessible_ranges_recursive(ranges, max_seq_length, ED, from, to);
>>  				}
>>  				else{
>> -					ranges.push_back(IndexRange(from+1,to+1));
>> +					ranges.push_back(IndexRange(from,to));
>>
>> huuh.. the index-start-1-convention only applies to user input!
>> internally you should use index-start-0 of course.. uaks.. sorry for the
>> confusion!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/BackofenLab/IntaRNA/pull/95#pullrequestreview-74709159>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AbAQ8ncmlihReexG3NRj0voi7YfbrEJnks5s0DhvgaJpZM4QRG1n>
>> .
>>
>
>

martin-mann commented:
With PR #96 the --qRange/--tRange parameter setup and range usage in IntaRNA.cpp gets obsolete. I now apply the subset definition directly during the parsing and thus avoid the storage of the obsolete sequences, which might be useful when dealing with really large files.

@agawrons thanks for the idea! It is already part of the `intarna-2` branch and will come online with v2.1.1

martin-mann commented:
I am still thinking about your ED-based decomposition.. thanks for the last output examples for `l=2` and `l=50`.. I guess I finally got your workflow..

what `RegionLenMax` did you apply? I guess something fairly large, since the reported regions are up 1300nt in length (l=50 `# INFO : 4750-6072`)

so to summarize, I think you do the following:

```
maxLength=1500; # some user-defined value
R = { 1-SEQLENGTH }; # range list
while (\exists r \in R : |r| > maxLength) do
- pick r \in R with |r| > maxLength
- find \argmax_{i \in r}( ED(i,i+l) ); # ie. position with max ED (interval length l)
- update R = R \setminus {r} \cup { r.from-i, (i+1)-r.to }
endWhile
```

which looks in general fine to me. The only critical point is, as discussed already, the choice of `l` used to compute the ED value for a cut point `i`.

using some large `l` is as arbitrary as using `l=1`. I would not use the seed-length since this triggers wrong intuitions. There is no connection between seed and this splitting. the only connection: you can skip a whole region if the minED(seedlength) is above a certain threshold (since all possible seeds will have positive energies and thus will be ignored), but I guess this is too much preprocessing and will be considered later anyway.. so here are options I see

1. `l=1`: 
  - *pro*: easy to understand
  - *con*: might not be optimal
2. `l` via parameter: 
  - *pro*: full control
  - *con*: additional parameter rarely used/changed by anybody, which complicates IntaRNA coding
3. `l` averaged for all windows of length <=XYZ spanning a position
  - *pro*: more general idea about accessibility of the region around i
  - *con*: computationally more expensive since increasing number of windows (number = 1+2+3+..+maxlength), but for small maxlength it should be fine and updateable! be position i linear number (2*maxlength) updates needed to get value from ED(i-1).
  - *con*: might also need another user parameter... (see (2.))

@agawrons what do you think? Given your experiences and tests, what makes sense?

martin-mann commented:
Final thinking on my way home brought me to the following conclusion:

Lets integrate your approach (in a non-recursive version) using `l=seedLength` and 
```
ED(i)=min_{wins of length l around i} ED(win)
```
ie. the considered ED for a cut pos  is the minimum over all seed-lengthed windows covering the position

This can be computed with a single update while screening for i and follows the central intarna idwa that a seed interaction should be present. Furthermore it avoids artifacts of single position ED values..

Any comments?

martin-mann commented:
ok, I implemented a non-recursive variant of your approach in ba36cbf7bb533e215aa8d18d3ad19a7c6230a69a

thus, I identify the ED window with maxED and removed the whole window (uses length = seedBP). this is iterated until all ranges are below a maximal length. I also enable and check a minimal range length (defaults to seedBP). furthermore, I renamed the argument to `--qRegionLenMax` to stay conform with other arguments.

I will add  according documentation to the README and merge the changes asap. Will be part of the next release.

@agawrons could you please verify/doublecheck that this is doing what you intend/expect? to this end, you can check-out the `accRegionDecomp` branch of the repository. Thanks!

martin-mann commented:
docu reads now:

If you are dealing with very long sequences it might be useful to use the
*automatic identification of accessible regions*, which dramatically reduces
runtime and memory consumption of IntaRNA since predictions are only done for
individual regions and not for the whole sequence. Here, we use a 
heuristic approach that finds and ignores subregions that are unlikely to form
an interaction, resulting in a decomposition of the full sequence range into
intervals of accessible regions. It can be enabled by providing the maximal 
length of the resulting intervals via the parameters `--qRegionLenMax` and
`--tRegionLenMax`.<br />
More specifically, starting from the full 
sequence's index range, the algorithm iteratively identifies in every too-long
range the window with highest ED value (penalty for non-accessibility). To
this end, it uses windows of length `--seedBP` to find subsequences where it is
most unlikely that a seed might be formed. This window is removed from the range,
which results in two shorter ranges. If a range is shorter than `--seedBP`, it
is completely removed.

agawrons commented:
Hello Martin,

Thanks for this! Non-recursive is usually better.

I tried to test your code with my test case and got this error:

# WARNING : Exception raised for :  #thread 0 #target 0 #query 0 :
STOP DUE TO MISSING IMPLEMENTATION : IndexRangeList::push_back() :
overlapping ranges are currently not supported
# WARNING : Exception raised : std::exception

I attached my input file. Parameters: --tAccL 150 --tAccW 200 --qAccL 150
--qAccW 200 -n 4 -q PCAT1.fa -t BRCA2_refseq.fa --tRegionLenMax 1500


On Mon, Nov 13, 2017 at 8:01 AM, Martin Raden <notifications@github.com>
wrote:

> Closed #95 <https://github.com/BackofenLab/IntaRNA/pull/95>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BackofenLab/IntaRNA/pull/95#event-1339012420>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AbAQ8lvMPuJEeVy9jNnQvY0MLL9mswclks5s2GfHgaJpZM4QRG1n>
> .
>

martin-mann commented:
the final version was now merged via #101 

thanks again for the contribution!
